File: 2022/13.js
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })

const check = require('./13-check.js')

const lines = []

let left = null
let right = null
let idx = 1

rl.on('line', data => {
  if (data === '') {
    lines.push({ left, right, idx, correct: false })
    left = null
    right = null
    idx += 1
    return
  }

  if (!left) {
    left = left || eval(data)
    return
  }
  if (!right) {
    right = right || eval(data)
  }
})

rl.on('close', () => {
  console.log('partOne', partOne(lines))
  console.log('partTwo', partTwo(lines))
})

function partOne (pairs) {
  for (const pair of pairs) {
    const { left, right } = pair
    pair.correct = check(left, right)
  }
  return pairs.filter(p => p.correct).reduce((acc, pair) => acc + pair.idx, 0)
}

function partTwo (pairs) {
  const divider2 = [[2]]
  const divider6 = [[6]]

  const packets = [divider2, divider6]
  for (const pair of pairs) {
    const { left, right } = pair
    packets.push(left)
    packets.push(right)
  }

  packets.sort((a, b) => check(a, b) ? -1 : 1)

  return (1 + packets.indexOf(divider2)) * (1 + packets.indexOf(divider6))
}

File: 2022/13.py
import sys

import functools

def compare(left, right):
    if isinstance(left, int) and isinstance(right, int):
        return left - right

    if isinstance(left, list) and isinstance(right, list):
        for l, r in zip(left, right):
            result = compare(l, r)
            if result:
                return result

        return len(left) - len(right)

    # exactly one value is integer
    if isinstance(left, int):
        return compare([left], right)

    return compare(left, [right])


def part_one(pairs):
    idxs = [i for i, (left, right) in enumerate(pairs, 1) if compare(eval(left), eval(right)) < 0]

    return sum(idxs)  # expect 5659

def part_two(pairs):
    divider2 = [[2]]
    divider6 = [[6]]

    packets = [divider2, divider6]
    for left, right in pairs:
        packets.append(eval(left))
        packets.append(eval(right))

    packets.sort(key=functools.cmp_to_key(compare))

    return (1 + packets.index(divider2)) * (1 + packets.index(divider6))


def main():
    lines = []
    left = None
    right = None
    for line in sys.stdin:
        line = line.replace('\n', '')
        if line == '':
            lines.append(( left, right ))
            left = None
            right = None
            continue

        if not left:
            left = line
            continue

        if not right:
            right = line
            continue

    if left and right:
        lines.append((left,right))

    print('part_one', part_one(lines))

    print('part_two', part_two(lines))


main()
